src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/TextEncoder.cs

```diff
@@ -322,6 +324,299 @@ public virtual void Encode(TextWriter output, char[] value, int startIndex, int
+                // data).
+
+                source = source.Slice(destination.Length);
+                if (!source.IsEmpty && UnicodeUtility.IsHighSurrogateCodePoint(source[source.Length - 1]))
+                {
+                    source = source.Slice(0, source.Length - 1);
+                }
+
+                source.CopyTo(destination); // guaranteed not to fail since we sliced earlier
+                charsConsumed = source.Length;
+                charsWritten = source.Length;
+                return OperationStatus.DestinationTooSmall;
+            }
+
+            // If we copied over all of the input data, success!
+
+            if (numCharsToCopy == source.Length)
+            {
+                charsConsumed = numCharsToCopy;
+                charsWritten = numCharsToCopy;
+                return OperationStatus.Done;
+            }
+
+            // There's data that must be encoded. Fall back to the scalar-by-scalar slow path.
+
+            int originalSourceLength = source.Length;
+            int originalDestinationLength = destination.Length;
+
+            source = source.Slice(numCharsToCopy);
+            destination = destination.Slice(numCharsToCopy);
+
+            while (!source.IsEmpty)
+            {
+                OperationStatus opStatus = UnicodeHelpers.DecodeScalarValueFromUtf16(source, out uint nextScalarValue, out int charsConsumedThisIteration);
+
+                switch (opStatus)
+                {
+                    case OperationStatus.Done:
+
+                        if (WillEncode((int)nextScalarValue))
+                        {
+                            goto default; // source data must be transcoded
+                        }
+                        else
+                        {
+                            // Source data can be copied as-is. Attempt to memcpy it to the destination buffer.
+
+                            if (source.Slice(0, charsConsumedThisIteration).TryCopyTo(destination))
+                            {
+                                destination = destination.Slice(charsConsumedThisIteration);
+                            }
+                            else
+                            {
+                                goto ReturnDestinationTooSmall;
+                            }
+                        }
+
+                        break;
+
+                    case OperationStatus.NeedMoreData:
+
+                        if (isFinalBlock)
+                        {
+                            goto default; // treat this as a normal invalid subsequence
+                        }
+                        else
+                        {
+                            goto ReturnNeedMoreData;
+                        }
+
+                    default:
+
+                        // This code path is hit for ill-formed input data (where decoding has replaced it with U+FFFD)
+                        // and for well-formed input data that must be escaped.
```

Comment:
We never return `OperationStatus.InvalidData` and I think we should rather than writin the replacement character.

I would rely on that to throw an exception:
https://github.com/dotnet/corefx/blob/c539d6c627b169d45f0b4cf1826b560cd0862abe/src/System.Text.Json/src/System/Text/Json/Writer/JsonWriterHelper.Escaping.cs#L115-L117

Reply:
All other `Encode*` APIs on this type substitute U+FFFD when given bad input. I don't think it'd be appropriate for us to change the behavior of newly introduced API given the existing behavior.

We could introduce a flag to control this behavior, similar to what we did for `Utf8.FromUtf16` and similar. (The Boolean parameter controls whether substitution is automatically performed vs. whether explicit failure is returned.) Do you have a scenario for this?

Reply:
> Do you have a scenario for this?

I want to detect that someone passed in invalid text and throw, not write replaced data.
```C#
using var writer = new Utf8JsonWriter(...);
writer.WriteStartObject();
// This should throw
writer.WriteStartArray(ReadOnlySpan<byte> containing invalid UTF-8 bytes);
writer.WriteEndArray();
writer.WriteEndObject();
```

> All other `Encode*` APIs on this type substitute U+FFFD when given bad input. I don't think it'd be appropriate for us to change the behavior of newly introduced API given the existing behavior.

Sure, except this one returns an `OperationStatus`. Shouldn't the default here be to not replace? 

> We could introduce a flag to control this behavior, similar to what we did for `Utf8.FromUtf16` and similar. (The Boolean parameter controls whether substitution is automatically performed vs. whether explicit failure is returned.)

It seems odd to provide a flag for a behavior that operation status already allows for, but I would be fine with it if you think that fits better with the existing APIs (especially since it is consistent with the `Utf8.FromUtf16` API).

Reply:
Per our offline conversation, I'm not blocking this for now. We can consider after this PR is complete.